File: test/sample.py
# Poikile Theme — Python Test File
#
# Colors vary by variant — see PALETTE.md for hex values per theme.
#
# Scopes to verify:
#   keyword.control.python         → keyword  (if, elif, else, for, while, try, except, finally, with, as, return, yield, raise, pass, break, continue, assert, del)
#   keyword.control.import.python  → keyword  (import, from)
#   keyword.operator.logical.python → keyword (and, or, not, in, is)
#   keyword.operator.unpacking.python → keyword (*, **)
#   storage.type.function.python   → keyword  (def, lambda)
#   storage.type.class.python      → keyword  (class)
#   storage.modifier.python        → keyword  italic (async, await)
#   entity.name.function.python    → function  (function names)
#   support.function.builtin.python → function (print, len, range, type, isinstance, etc.)
#   entity.name.type.class.python  → type  (class names)
#   support.type.python            → type  (int, str, float, bool, list, dict, tuple, set, bytes, type)
#   entity.name.function.decorator.python → decorator italic (@decorator)
#   punctuation.definition.decorator.python → decorator italic
#   variable.language.python       → keyword  italic (self, cls)
#   variable.parameter.python      → fg.default  (parameters)
#   constant.language.python       → number  (True, False, None)
#   support.variable.magic.python  → decorator  (__init__, __name__, etc.)
#   storage.type.string.python     → regex  (f, r, b, u string prefixes)
#   meta.fstring.python            → regex
#   string.interpolated.python     → regex
#   meta.type.annotation.python    → type  (type hints)
#   string                         → string  (strings)
#   constant.numeric               → number  (numbers)
#   comment                        → fg.muted  italic

from __future__ import annotations

import asyncio
import json
import logging
import re
from abc import ABC, abstractmethod
from collections.abc import AsyncIterator, Callable, Sequence
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum, auto
from functools import wraps
from pathlib import Path
from typing import (
    Any,
    ClassVar,
    Final,
    Generic,
    Literal,
    Protocol,
    TypeAlias,
    TypeVar,
    overload,
    runtime_checkable,
)

logger = logging.getLogger(__name__)

# ── Type Variables & Aliases ─────────────────────────────────────────────

T = TypeVar("T")
E = TypeVar("E", bound=Exception)
JsonDict: TypeAlias = dict[str, Any]
Callback: TypeAlias = Callable[..., Any]

# ── Enums ────────────────────────────────────────────────────────────────


class Priority(Enum):
    LOW = auto()
    MEDIUM = auto()
    HIGH = auto()
    CRITICAL = auto()


class TaskStatus(Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


# ── Decorators ───────────────────────────────────────────────────────────


def retry(max_attempts: int = 3, delay: float = 1.0):
    """Decorator that retries a function on failure with exponential backoff."""

    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        async def async_wrapper(*args: Any, **kwargs: Any) -> T:
            last_error: Exception | None = None
            for attempt in range(max_attempts):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    last_error = e
                    wait = delay * (2**attempt)
                    logger.warning(
                        f"Attempt {attempt + 1}/{max_attempts} failed: {e}. "
                        f"Retrying in {wait:.1f}s..."
                    )
                    await asyncio.sleep(wait)
            raise last_error  # type: ignore[misc]

        @wraps(func)
        def sync_wrapper(*args: Any, **kwargs: Any) -> T:
            last_error: Exception | None = None
            for attempt in range(max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_error = e
                    import time

                    time.sleep(delay * (2**attempt))
            raise last_error  # type: ignore[misc]

        if asyncio.iscoroutinefunction(func):
            return async_wrapper  # type: ignore[return-value]
        return sync_wrapper  # type: ignore[return-value]

    return decorator


def validate_positive(func: Callable[..., T]) -> Callable[..., T]:
    """Ensure all numeric arguments are positive."""

    @wraps(func)
    def wrapper(*args: Any, **kwargs: Any) -> T:
        for arg in args:
            if isinstance(arg, (int, float)) and arg < 0:
                raise ValueError(f"Expected positive number, got {arg}")
        for key, val in kwargs.items():
            if isinstance(val, (int, float)) and val < 0:
                raise ValueError(f"Parameter '{key}' must be positive, got {val}")
        return func(*args, **kwargs)

    return wrapper  # type: ignore[return-value]


# ── Protocols ────────────────────────────────────────────────────────────


@runtime_checkable
class Serializable(Protocol):
    def to_dict(self) -> JsonDict: ...
    @classmethod
    def from_dict(cls, data: JsonDict) -> "Serializable": ...


class Repository(Protocol[T]):
    async def get(self, id: str) -> T | None: ...
    async def save(self, item: T) -> None: ...
    async def delete(self, id: str) -> bool: ...
    async def list_all(self, limit: int = 100, offset: int = 0) -> list[T]: ...


# ── Data Classes ─────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class TaskConfig:
    timeout: float = 30.0
    max_retries: int = 3
    priority: Priority = Priority.MEDIUM
    tags: frozenset[str] = frozenset()

    def with_priority(self, priority: Priority) -> TaskConfig:
        return TaskConfig(
            timeout=self.timeout,
            max_retries=self.max_retries,
            priority=priority,
            tags=self.tags,
        )


@dataclass
class Task:
    id: str
    title: str
    description: str | None = None
    status: TaskStatus = TaskStatus.PENDING
    config: TaskConfig = field(default_factory=TaskConfig)
    created_at: datetime = field(default_factory=datetime.now)
    completed_at: datetime | None = None
    _attempt_count: int = field(default=0, repr=False)

    # Class variable shared by all instances
    MAX_TITLE_LENGTH: ClassVar[int] = 200

    def __post_init__(self) -> None:
        if len(self.title) > self.MAX_TITLE_LENGTH:
            raise ValueError(
                f"Title exceeds {self.MAX_TITLE_LENGTH} chars: "
                f"'{self.title[:50]}...'"
            )

    @property
    def is_terminal(self) -> bool:
        return self.status in (TaskStatus.COMPLETED, TaskStatus.FAILED, TaskStatus.CANCELLED)

    @property
    def duration(self) -> timedelta | None:
        if self.completed_at is None:
            return None
        return self.completed_at - self.created_at

    def to_dict(self) -> JsonDict:
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "status": self.status.value,
            "priority": self.config.priority.name.lower(),
            "created_at": self.created_at.isoformat(),
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "tags": sorted(self.config.tags),
        }

    @classmethod
    def from_dict(cls, data: JsonDict) -> Task:
        config = TaskConfig(
            priority=Priority[data.get("priority", "medium").upper()],
            tags=frozenset(data.get("tags", [])),
        )
        return cls(
            id=data["id"],
            title=data["title"],
            description=data.get("description"),
            status=TaskStatus(data.get("status", "pending")),
            config=config,
            created_at=datetime.fromisoformat(data["created_at"]),
        )


# ── Abstract Base Class ─────────────────────────────────────────────────


class BaseExecutor(ABC, Generic[T]):
    """Abstract base for task executors with lifecycle hooks."""

    REGISTRY: ClassVar[dict[str, type["BaseExecutor[Any]"]]] = {}

    def __init_subclass__(cls, executor_type: str | None = None, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        if executor_type is not None:
            BaseExecutor.REGISTRY[executor_type] = cls

    @abstractmethod
    async def execute(self, task: Task) -> T:
        """Execute the task and return a result."""
        ...

    @abstractmethod
    async def rollback(self, task: Task, error: Exception) -> None:
        """Handle cleanup after a failed execution."""
        ...

    async def before_execute(self, task: Task) -> None:
        logger.info(f"Starting execution of task {task.id}")

    async def after_execute(self, task: Task, result: T) -> None:
        logger.info(f"Completed task {task.id}")


# ── Concrete Implementation ─────────────────────────────────────────────


class HttpExecutor(BaseExecutor[bytes], executor_type="http"):

    def __init__(self, base_url: str, *, verify_ssl: bool = True) -> None:
        self.base_url = base_url.rstrip("/")
        self.verify_ssl = verify_ssl
        self._session_count = 0

    @retry(max_attempts=3, delay=0.5)
    async def execute(self, task: Task) -> bytes:
        import aiohttp

        url = f"{self.base_url}/tasks/{task.id}/execute"
        timeout = aiohttp.ClientTimeout(total=task.config.timeout)

        async with aiohttp.ClientSession(timeout=timeout) as session:
            self._session_count += 1
            async with session.post(url, json=task.to_dict()) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise RuntimeError(
                        f"HTTP {resp.status} from {url}: {body[:200]}"
                    )
                return await resp.read()

    async def rollback(self, task: Task, error: Exception) -> None:
        logger.error(f"Rolling back task {task.id}: {error}")


# ── Async Generator ─────────────────────────────────────────────────────


async def task_stream(
    tasks: Sequence[Task],
    *,
    batch_size: int = 10,
) -> AsyncIterator[tuple[Task, TaskStatus]]:
    """Yield tasks in batches with status updates."""
    for i in range(0, len(tasks), batch_size):
        batch = tasks[i : i + batch_size]
        results = await asyncio.gather(
            *[_process_single(t) for t in batch],
            return_exceptions=True,
        )
        for task, result in zip(batch, results):
            if isinstance(result, Exception):
                yield task, TaskStatus.FAILED
            else:
                yield task, TaskStatus.COMPLETED


async def _process_single(task: Task) -> None:
    await asyncio.sleep(0.01)


# ── Overloaded Functions ─────────────────────────────────────────────────


@overload
def parse_value(raw: str, as_type: Literal["int"]) -> int: ...
@overload
def parse_value(raw: str, as_type: Literal["float"]) -> float: ...
@overload
def parse_value(raw: str, as_type: Literal["bool"]) -> bool: ...
@overload
def parse_value(raw: str, as_type: Literal["str"]) -> str: ...


def parse_value(raw: str, as_type: str) -> int | float | bool | str:
    match as_type:
        case "int":
            return int(raw)
        case "float":
            return float(raw)
        case "bool":
            return raw.lower() in ("true", "1", "yes")
        case "str":
            return raw
        case _:
            raise ValueError(f"Unknown type: {as_type}")


# ── Context Manager ──────────────────────────────────────────────────────


class Timer:
    """Measure elapsed time as a context manager."""

    def __init__(self, label: str = "operation") -> None:
        self.label = label
        self.elapsed: float = 0.0
        self._start: float = 0.0

    def __enter__(self) -> "Timer":
        import time

        self._start = time.perf_counter()
        return self

    def __exit__(self, *exc: Any) -> None:
        import time

        self.elapsed = time.perf_counter() - self._start
        logger.info(f"{self.label} took {self.elapsed:.4f}s")


# ── String & Regex Patterns ──────────────────────────────────────────────

VERSION_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"^v?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)"
    r"(?:-(?P<pre>[a-zA-Z0-9.]+))?"
    r"(?:\+(?P<build>[a-zA-Z0-9.]+))?$"
)

MULTILINE_SQL = """
    SELECT t.id, t.title, t.status, c.priority
    FROM tasks t
    JOIN task_config c ON c.task_id = t.id
    WHERE t.status = 'pending'
      AND c.priority IN ('high', 'critical')
    ORDER BY t.created_at DESC
    LIMIT %(limit)s OFFSET %(offset)s
"""

RAW_PATH = r"C:\Users\admin\Documents\file.txt"
BYTE_STRING = b"\x00\x01\x02\xff"

# ── Numeric Literals ─────────────────────────────────────────────────────

PI: Final[float] = 3.14159_26535
HEX_COLOR: Final[int] = 0xFF_AA_00
OCTAL_PERMS: Final[int] = 0o755
BINARY_FLAGS: Final[int] = 0b1010_0101
BIG_NUMBER: Final[int] = 1_000_000_000
SCIENTIFIC: Final[float] = 6.022e23
NEGATIVE_EXP: Final[float] = 1.6e-19
COMPLEX_NUM: Final[complex] = 3 + 4j

# ── Main ─────────────────────────────────────────────────────────────────


async def main() -> None:
    config = TaskConfig(
        timeout=15.0,
        max_retries=5,
        priority=Priority.HIGH,
        tags=frozenset({"deploy", "production"}),
    )

    tasks = [
        Task(id=f"task-{i:03d}", title=f"Process batch #{i}", config=config)
        for i in range(1, 21)
    ]

    with Timer("batch processing"):
        async for task, status in task_stream(tasks, batch_size=5):
            symbol = "✓" if status == TaskStatus.COMPLETED else "✗"
            print(f"  {symbol} {task.id}: {status.value}")

    # Demonstrate various builtins
    lengths = list(map(len, [t.title for t in tasks]))
    total = sum(lengths)
    avg = total / len(lengths) if lengths else 0
    sorted_tasks = sorted(tasks, key=lambda t: t.title, reverse=True)
    unique_statuses = set(t.status for t in tasks)
    task_dict = {t.id: t.to_dict() for t in tasks[:3]}

    print(f"\nProcessed {len(tasks)} tasks")
    print(f"Average title length: {avg:.1f}")
    print(f"Unique statuses: {unique_statuses}")
    print(f"First 3 as JSON:\n{json.dumps(task_dict, indent=2)}")

    # Type checks
    assert isinstance(tasks[0], Task)
    assert not isinstance(tasks[0], str)
    assert issubclass(HttpExecutor, BaseExecutor)

    version_match = VERSION_PATTERN.match("v2.1.0-beta.1+build.42")
    if version_match is not None:
        print(f"Version: {version_match.group('major')}.{version_match.group('minor')}")

    value: int | str = parse_value("42", "int")
    assert value == 42 and type(value) is int

    # Walrus operator and comprehensions
    if (n := len(tasks)) > 10:
        heavy = [t for t in tasks if t.config.priority in (Priority.HIGH, Priority.CRITICAL)]
        print(f"{n} tasks total, {len(heavy)} are high priority")

    # Unpacking
    first, second, *rest = tasks
    print(f"First: {first.id}, remaining: {len(rest)}")


if __name__ == "__main__":
    asyncio.run(main())

File: test/sample.rb
# Poikile Theme — Ruby Test File
#
# Scopes to verify:
#   keyword.control.ruby                       → #c2788e  (if, elsif, else, unless, while, until, for, do, end, begin, rescue, ensure, raise, return, yield, case, when, then, class, module, def)
#   storage.type.ruby                          → #c2788e  (def, class, module)
#   entity.name.function.ruby                  → #d4a55c  (method names)
#   entity.name.type.class.ruby                → #7db89e  (class names)
#   entity.name.type.module.ruby               → #7db89e  (module names)
#   support.function.kernel.ruby               → #d4a55c  (puts, gets, require, raise, attr_accessor, attr_reader, attr_writer)
#   variable.language.ruby                     → #c2788e  italic (self)
#   variable.other.readwrite.instance.ruby     → #d4d0c8  (@instance)
#   variable.other.readwrite.class.ruby        → #d4d0c8  (@@class)
#   variable.other.readwrite.global.ruby       → #d4d0c8  ($global)
#   punctuation.definition.variable.ruby       → #d4d0c8
#   constant.other.symbol.ruby                 → #c8ba5e  (:symbol)
#   string.interpolated.ruby                   → #d08a6e  (#{interpolation})
#   constant.language.ruby                     → #c9905a  (true, false, nil)
#   keyword.operator.ruby                      → #9b9baa
#   string                                     → #a3b87c
#   constant.numeric                           → #c9905a
#   comment                                    → #8a8784  italic

require 'json'
require 'net/http'
require 'uri'
require 'logger'
require 'securerandom'
require 'date'

# ── Module for shared behavior ───────────────────────────────────────────

module Serializable
  def to_json(*args)
    to_h.to_json(*args)
  end

  def to_h
    raise NotImplementedError, "#{self.class} must implement #to_h"
  end
end

module Describable
  def describe
    raise NotImplementedError, "#{self.class} must implement #describe"
  end

  def summary
    desc = describe
    desc.length > 80 ? "#{desc[0..76]}..." : desc
  end
end

# ── Enum-like module ─────────────────────────────────────────────────────

module Priority
  LOW = :low
  MEDIUM = :medium
  HIGH = :high
  CRITICAL = :critical

  ALL = [LOW, MEDIUM, HIGH, CRITICAL].freeze

  def self.valid?(value)
    ALL.include?(value)
  end

  def self.from_string(str)
    sym = str.to_s.downcase.to_sym
    raise ArgumentError, "Invalid priority: #{str}" unless valid?(sym)
    sym
  end
end

# ── Configuration with class variables ───────────────────────────────────

class Config
  @@instances = 0

  attr_accessor :base_url, :timeout, :max_retries, :verify_ssl
  attr_reader :created_at

  DEFAULT_TIMEOUT = 30
  MAX_RETRY_COUNT = 5

  def initialize(base_url:, timeout: DEFAULT_TIMEOUT, max_retries: 3, verify_ssl: true)
    @base_url = base_url.chomp('/')
    @timeout = timeout
    @max_retries = [max_retries, MAX_RETRY_COUNT].min
    @verify_ssl = verify_ssl
    @created_at = Time.now
    @@instances += 1
  end

  def self.instance_count
    @@instances
  end

  def to_s
    "#<Config url=#{@base_url} timeout=#{@timeout}s retries=#{@max_retries}>"
  end
end

# ── Task class with full Ruby features ───────────────────────────────────

class Task
  include Serializable
  include Describable
  include Comparable

  attr_accessor :title, :description, :status, :priority
  attr_reader :id, :created_at, :metadata

  STATUSES = %i[pending running completed failed cancelled].freeze

  def initialize(id: nil, title:, description: nil, priority: Priority::MEDIUM)
    @id = id || SecureRandom.uuid
    @title = title
    @description = description
    @priority = priority
    @status = :pending
    @metadata = {}
    @created_at = Time.now
    @completed_at = nil

    yield self if block_given?
  end

  def complete!
    raise "Task already terminal" if terminal?
    @status = :completed
    @completed_at = Time.now
    self
  end

  def fail!(reason = nil)
    raise "Task already terminal" if terminal?
    @status = :failed
    @metadata[:failure_reason] = reason if reason
    self
  end

  def terminal?
    %i[completed failed cancelled].include?(@status)
  end

  def duration
    return nil unless @completed_at
    @completed_at - @created_at
  end

  def add_metadata(key, value)
    @metadata[key.to_sym] = value
    self
  end

  def <=>(other)
    return nil unless other.is_a?(Task)
    Priority::ALL.index(priority) <=> Priority::ALL.index(other.priority)
  end

  def to_h
    {
      id: @id,
      title: @title,
      description: @description,
      status: @status,
      priority: @priority,
      created_at: @created_at.iso8601,
      metadata: @metadata
    }
  end

  def describe
    "[#{@id[0..7]}] #{@title} (#{@priority}, #{@status})"
  end

  def to_s
    "Task(#{@id[0..7]}, \"#{@title}\")"
  end

  def inspect
    "#<Task id=#{@id[0..7]} title=#{@title.inspect} status=#{@status}>"
  end
end

# ── Repository with error handling ───────────────────────────────────────

class TaskRepository
  class NotFoundError < StandardError; end
  class ValidationError < StandardError; end

  def initialize
    @store = {}
    @logger = Logger.new($stdout)
    @logger.level = Logger::INFO
  end

  def save(task)
    raise ValidationError, "Title cannot be blank" if task.title.nil? || task.title.strip.empty?
    @store[task.id] = task
    @logger.info("Saved task: #{task.id}")
    task
  end

  def find(id)
    @store[id] or raise NotFoundError, "Task not found: #{id}"
  end

  def find_by_status(status)
    @store.values.select { |t| t.status == status }
  end

  def find_by_priority(*priorities)
    @store.values.select { |t| priorities.include?(t.priority) }
  end

  def all
    @store.values.sort
  end

  def delete(id)
    task = @store.delete(id)
    raise NotFoundError, "Task not found: #{id}" unless task
    @logger.info("Deleted task: #{id}")
    task
  end

  def count
    @store.size
  end

  def each(&block)
    @store.values.each(&block)
  end

  def map(&block)
    @store.values.map(&block)
  end
end

# ── HTTP client with retry logic ─────────────────────────────────────────

class HttpClient
  MAX_REDIRECTS = 5

  def initialize(config)
    @config = config
    @logger = Logger.new($stdout)
  end

  def get(path, headers: {})
    request(:get, path, headers: headers)
  end

  def post(path, body:, headers: {})
    request(:post, path, body: body, headers: headers)
  end

  private

  def request(method, path, body: nil, headers: {})
    url = URI.parse("#{@config.base_url}#{path}")
    attempts = 0

    begin
      attempts += 1
      http = Net::HTTP.new(url.host, url.port)
      http.use_ssl = url.scheme == 'https'
      http.verify_mode = @config.verify_ssl ? OpenSSL::SSL::VERIFY_PEER : OpenSSL::SSL::VERIFY_NONE
      http.read_timeout = @config.timeout

      req = case method
            when :get then Net::HTTP::Get.new(url)
            when :post
              r = Net::HTTP::Post.new(url)
              r.body = body.to_json if body
              r
            else
              raise ArgumentError, "Unsupported method: #{method}"
            end

      headers.each { |k, v| req[k.to_s] = v.to_s }
      req['Content-Type'] = 'application/json'
      req['Accept'] = 'application/json'

      response = http.request(req)

      case response
      when Net::HTTPSuccess
        JSON.parse(response.body, symbolize_names: true)
      when Net::HTTPRedirection
        raise "Too many redirects" if attempts > MAX_REDIRECTS
        url = URI.parse(response['location'])
        retry
      else
        raise "HTTP #{response.code}: #{response.message}"
      end
    rescue StandardError => e
      if attempts < @config.max_retries
        delay = 2**attempts
        @logger.warn("Attempt #{attempts} failed: #{e.message}. Retrying in #{delay}s...")
        sleep(delay)
        retry
      end
      raise
    end
  end
end

# ── Blocks, procs, lambdas ───────────────────────────────────────────────

def with_timing(label = "operation")
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  result = yield
  elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
  puts "#{label} took #{elapsed.round(4)}s"
  result
end

def transform_tasks(tasks, &block)
  tasks.map(&block)
end

square = ->(x) { x * x }
double = proc { |x| x * 2 }

# ── String operations ────────────────────────────────────────────────────

def format_report(tasks)
  header = <<~HEREDOC
    ╔═══════════════════════════════════╗
    ║       Task Pipeline Report        ║
    ╚═══════════════════════════════════╝
  HEREDOC

  rows = tasks.map do |task|
    status_icon = case task.status
                  when :completed then '✓'
                  when :failed then '✗'
                  when :running then '⟳'
                  else '○'
                  end
    "  #{status_icon} #{task.describe}"
  end

  "#{header}\n#{rows.join("\n")}\n\nTotal: #{tasks.size} tasks"
end

# ── Numeric literals ─────────────────────────────────────────────────────

PI = 3.14159_26535
HEX_COLOR = 0xFF_AA_00
OCTAL_PERMS = 0o755
BINARY_FLAGS = 0b1010_0101
BIG_NUMBER = 1_000_000_000
RATIONAL = 1/3r
COMPLEX = 3 + 4i

# ── Main execution ───────────────────────────────────────────────────────

if __FILE__ == $PROGRAM_NAME
  repo = TaskRepository.new

  tasks = [
    Task.new(title: "Deploy service", priority: Priority::HIGH),
    Task.new(title: "Run migrations", priority: Priority::CRITICAL),
    Task.new(title: "Update docs", priority: Priority::LOW),
    Task.new(title: "Send notifications") { |t| t.add_metadata(:channel, "slack") },
  ]

  tasks.each { |task| repo.save(task) }

  with_timing("task processing") do
    repo.each do |task|
      begin
        task.complete!
        puts "Completed: #{task}"
      rescue => e
        puts "Error: #{e.message}"
        task.fail!(e.message)
      ensure
        puts "  Status: #{task.status}"
      end
    end
  end

  # Enumerable operations
  high_priority = repo.find_by_priority(Priority::HIGH, Priority::CRITICAL)
  completed = repo.find_by_status(:completed)
  titles = repo.map(&:title)
  descriptions = transform_tasks(repo.all) { |t| t.summary }

  puts "\nHigh priority: #{high_priority.size}"
  puts "Completed: #{completed.size}"
  puts "Titles: #{titles.inspect}"
  puts "Report:\n#{format_report(repo.all)}"

  # Demonstrate lambdas
  numbers = [1, 2, 3, 4, 5]
  puts "Squares: #{numbers.map(&square).inspect}"
  puts "Doubles: #{numbers.map(&double).inspect}"

  # Numeric operations
  puts "PI: #{PI}, hex: #{HEX_COLOR}, rational: #{RATIONAL.to_f}"

  # Regex
  version = "v2.1.0-beta.1"
  if version =~ /^v?(\d+)\.(\d+)\.(\d+)/
    puts "Version: #{$1}.#{$2}.#{$3}"
  end

  # Global variable
  $exit_code = 0
  exit($exit_code)
end
